### PR TITLE
py38: upgrade to DRF 3.8.x

### PR DESCRIPTION
**Pull request by joshuarli**
3.9.2 supports Django 2.2, but they add official 3.8 support in 3.11: https://github.com/encode/django-rest-framework/commit/de497a9bf12605b8b71bf7c21da57bc2a8238786 Reading the [changelog](https://www.django-rest-framework.org/community/release-notes/), 3.9 to 3.11 should be an easy bump, it's just 3.7.x and 3.8.x having some things that need attention.

I tested it on staging, it seems good. https://github.com/getsentry/getsentry/pull/5340

3.7.x:

The BC-incompatible changes (search for "BC Change" [here](https://www.django-rest-framework.org/community/release-notes/#37x-series)) don't apply to us (example: the DateTimeField change doesn't apply; we have USE_TZ=True and TIME_ZONE="utc") or are desirable (example: STRICT_JSON).

The one change that broke our tests was that they added a request param [here](https://github.com/encode/django-rest-framework/pull/5295/files) leading to:

```
Traceback (most recent call last):
  File "/Users/joshua.li/dev/sentry/sentry/src/sentry/api/base.py", line 119, in handle_exception
    response = super().handle_exception(exc)
  File "/Users/joshua.li/dev/sentry/sentry/.venv/lib/python3.6/site-packages/rest_framework/views.py", line 443, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/Users/joshua.li/dev/sentry/sentry/src/sentry/api/base.py", line 212, in dispatch
    self.initial(*args, request=request, **kwargs)
  File "/Users/joshua.li/dev/sentry/sentry/.venv/lib/python3.6/site-packages/sentry_sdk/integrations/django/__init__.py", line 264, in sentry_patched_drf_initial
    return old_drf_initial(self, request, *args, **kwargs)
  File "/Users/joshua.li/dev/sentry/sentry/.venv/lib/python3.6/site-packages/rest_framework/views.py", line 388, in initial
    self.perform_authentication(request)
  File "/Users/joshua.li/dev/sentry/sentry/.venv/lib/python3.6/site-packages/rest_framework/views.py", line 314, in perform_authentication
    request.user
  File "/Users/joshua.li/dev/sentry/sentry/.venv/lib/python3.6/site-packages/rest_framework/request.py", line 222, in user
    self._authenticate()
  File "/Users/joshua.li/dev/sentry/sentry/.venv/lib/python3.6/site-packages/rest_framework/request.py", line 375, in _authenticate
    user_auth_tuple = authenticator.authenticate(self)
  File "/Users/joshua.li/dev/sentry/sentry/.venv/lib/python3.6/site-packages/rest_framework/authentication.py", line 87, in authenticate
    return self.authenticate_credentials(userid, password, request)
TypeError: authenticate_credentials() takes 3 positional arguments but 4 were given
```

Side note, I checked sentry_patched_drf_initial with drf 3.8, it's fine.

We have to accept this in one of our authenticate_credentials:

```
> /Users/joshua.li/dev/sentry/sentry/.venv/lib/python3.6/site-packages/rest_framework/authentication.py(88)authenticate()
-> return self.authenticate_credentials(userid, password, request)
(Pdb) self.authenticate_credentials(userid, password, request)
*** TypeError: authenticate_credentials() takes 3 positional arguments but 4 were given
(Pdb) self.authenticate_credentials
<bound method ApiKeyAuthentication.authenticate_credentials of <sentry.api.authentication.ApiKeyAuthentication object at 0x10a628240>>
```

3.8.x:

The main thing that stood out to me in this release that applied to sentry was https://github.com/encode/django-rest-framework/pull/5655, which makes serializer validation errors better for composite fields (ListField, DictField) and as a consequence, fundamentally changes their structure.

This was visible in some test breakage that checks against things like serializer.errors and responses.data. Fixing serializer.errors isn't a problem (they also added `__eq__` for ErrorDetails to make this easier), but response.data... well first I refactored it to json, and that also led me to notice that in most 400 responses we just chuck in the serializer.errors. So breakage here would be any client that does error handling and assumes some particular structure for 400s, which is to say, unlikely. I also don't believe we guarantee any kind of schema for 400 responses, for example, https://docs.sentry.io/api/organizations/update-an-organization/ we just have the human meaning for non-2XX codes.

### CODE REVIEW

**Comment by joshuarli** [comment]
Oof okay, DRF 3.8.x is a little weird for us, I'm gonna do 3.7.x then 3.8.x then hopefully straight to 3.11.x.

File: src/sentry/api/authentication.py [hunk 1]
```diff
@@ -137,6 +137,7 @@ def authenticate(self, request):
 class TokenAuthentication(StandardAuthentication):
     token_name = b"bearer"
 
+    # might need to change this funcsig?
```
Comment (joshuarli) [inline src/sentry/api/authentication.py:140]: TODO need to test this.

Comment (joshuarli) [reply src/sentry/api/authentication.py:140]: Oh, just realized, it doesn't apply to `StandardAuthentication`.

**Comment by scefali** [comment]
I believe we do expect specific schema on 400 responses in some cases. I’ll dig up an example after I eat lunch.

**Comment by joshuarli** [comment]
> I believe we do expect specific schema on 400 responses in some cases. I’ll dig up an example after I eat lunch.

For ourselves somehow? Or do you mean guarantees?

**Comment by scefali** [comment]
> For ourselves somehow? Or do you mean guarantees that we make?

@joshuarli  For ourselves. With forms for example, we pull the error out of the response using the field name as the key so we can highlight that field in the form UI. Here's how that works:

https://github.com/getsentry/sentry/blob/2829e61be5cec1b9974e34c10021e5d95455221b/static/app/views/settings/components/forms/model.tsx#L561-L563

But I'm not sure if this change is impacting endpoints used in our forms.

**Comment by joshuarli** [comment]
> > For ourselves somehow? Or do you mean guarantees that we make?
> 
> @joshuarli For ourselves. With forms for example, we pull the error out of the response using the field name as the key so we can highlight that field in the form UI. Here's how that works:
> 
> https://github.com/getsentry/sentry/blob/2829e61be5cec1b9974e34c10021e5d95455221b/static/app/views/settings/components/forms/model.tsx#L561-L563
> 
> But I'm not sure if this change is impacting endpoints used in our forms.

Oh yeah, it'd definitely just go to `this.setError(id, 'Failed to save');` instead of a more specific error. Though, having a hard time searching for any user-supplied forms that use ListFields.

**Comment by joshuarli** [comment]
After some more discussion, I've decided the path of least resistance forward is to monkeypatch.

Maintenance burden won't be too bad because I'll also be upgrading to 3.11.x so I'm aware of what to check, and beyond that I don't see us upgrading DRF much in the near future.

File: tests/sentry/api/serializers/test_fields.py [hunk 2]
```diff
@@ -30,14 +31,18 @@ def test_allow_null(self):
 
         serializer = DummySerializer(data=data)
         assert not serializer.is_valid()
-        assert serializer.errors == {"a_field": ["This field may not be null."]}
+        assert serializer.errors == {
+            "a_field": [ErrorDetail(string="This field may not be null.", code="null")]
+        }
```
Comment (wedamija) [inline tests/sentry/api/serializers/test_fields.py:36]: These changes are just internal I assume, and we still produce the output via the api?

**Review by wedamija — APPROVED** [review]
Looks good, have you clicked around on staging with these prs?

File: tests/sentry/api/serializers/test_fields.py (hunk 2 above)
Comment (joshuarli) [reply tests/sentry/api/serializers/test_fields.py:36]: Yeah.

File: src/sentry/api/serializers/rest_framework/base.py [hunk 3]
```diff
@@ -25,11 +25,18 @@ def convert_dict_key_case(obj, converter):
...
 
     obj = obj.copy()
     for key in list(obj.keys()):
+        if isinstance(key, int):
+            # In DRF 3.8+, ListFields will have ints following
+            # https://github.com/encode/django-rest-framework/pull/5655.
+            # Don't convert the key in this case.
+            obj[key] = convert_dict_key_case(obj.pop(key), converter)
+            continue
         converted_key = converter(key)
```
[the remark this answers is not in the document]
Comment (joshuarli) [reply src/sentry/api/serializers/rest_framework/base.py]: You're referring to outdated code, that was only needed before monkeypatching.

[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/api/serializers/rest_framework/base.py]: Oops, I think I meant to remove this comment because I wrote it ages ago

**Comment by joshuarli** [comment]
> Looks good, have you clicked around on staging with these prs?

Yeah.

**Pull request merged**